Agent hint header implementation in .NET APIs

joelhulen opened the pull request:
# Agent hint header implementation in .NET APIs

<!-- Thank you for contributing to FoundationaLLM!  Open source is only as strong as its contributors. -->

## The issue or feature being addressed

N/A

## Details on the issue fix or feature implementation

Provides a common way to intercept and inject the agent hint HTTP header. Implemented the header middleware and context within `CoreAPI`, `GatekeeperAPI`, and `AgentFactoryAPI`. Created tests for the new middleware as well as existing `HttpClientFactoryService`.

Also cleaned up the Agent Factory and its related API service. The user context needs to flow down the request chain via the X-USER-CONTEXT header to the downstream Hub APIs. Passing the context any other way is incorrect.

## Confirm the following

- [x]  I started this PR by branching from the head of the default branch
- [x]  I have targeted the PR to merge into the default branch
- [x]  I have included unit tests for the issue/feature
- [x]  I have included inline docs for my changes, where applicable
- [x]  I have successfully run a local build

ciprianjichici reviewed (commented):
Instead of splitting the information into multiple data structures, I think it would make more sense to have an `IExecutionContext` of `ICallContext` which would have as properties the UPN and AgentHint (for now). 
Will make it significantly easier in the future to add new context properties (which we will most likely need to do anyway).

joelhulen commented:
> Instead of splitting the information into multiple data structures, I think it would make more sense to have an `IExecutionContext` of `ICallContext` which would have as properties the UPN and AgentHint (for now). Will make it significantly easier in the future to add new context properties (which we will most likely need to do anyway).

I'll make this change. I had considered doing that anyway.

joelhulen commented:
@ciprianjichici, I've addressed your feedback. Also, I had to merge in my Agent Factory cleanup branch into this one due to some of the changes I had to make. Because of this, I closed out my cleanup PR. That's why there are a lot of extra file changes in this PR now.
